Feat/backups upgrade path  -- WIP

ZenGround0 opened the pull request:
The basic approach is to read the json backup and replace all raft state with a single snapshot of the new version, while preserving the necessary raft config meta-data.  I added `migration` as an ipfs-cluster-service rather than ipfs-cluster-ctl command because it reinforces to the user that the ipfs-cluster-service daemon should be stopped, reuses some of the config functions already built into this tool, and can't talk to an ipfs-cluster-service daemon endpoint so including it in ipfs-cluster-ctl would break the common pattern.

- Still testing this out
- Needs sharness, possibly other tests
- Still thinking about upgrading general migration framework

--Update--
@hsanjuan The core functionality is here and tests out manually.  When you have a chance your feedback would be awesome.  This is still WIP as I want to add a few things (sharness, cleaner development upgrade path and UX for trying to load bad raft state after an update without a migration).  Thanks!

ZenGround0 commented on ipfs-cluster-service/main.go:
```diff
@@ -434,3 +531,42 @@ func makeConfigs() (*config.Manager, *ipfscluster.Config, *rest.Config, *ipfshtt
+	return bhost, nil
+}
+	
+func encodeState(state mapstate.MapState) ([]byte, error) {
```
This is just copy pasted from libp2p-raft/codec.  I'd like your opinion on whether this should be exported from the module or kept inline here.

ZenGround0 commented on util.go:
```diff
@@ -52,6 +52,10 @@ func copyEmptyStructToIfaces(in []struct{}) []interface{} {
 	return ifaces
 }
 
+func MultiaddrSplit(addr ma.Multiaddr) (peer.ID, ma.Multiaddr, error) {
```
I shadowed multiaddrSplit with an exported version but probably shouldn't keep it this way.  I wanted to check if you think it is OK to export.

ZenGround0 commented on ipfs-cluster-service/main.go:
```diff
@@ -231,6 +242,74 @@ configuration.
+					raftDataPath = c.Args().Get(1)
+				}
+				
+				//Migrate backup to new state
```
If you think there is a better file (maybe I should make one?) to execute these things I am open to suggestions.

hsanjuan replied:
Perhaps let's put most of this action in a `migrate()` function. `migrate.go` file won't hurt either.

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -182,6 +182,10 @@ func makeDataFolder(baseDir, folder string) (string, error) {
 	return folder, nil
 }
 
+func MakeServerConf(peers []peer.ID) hraft.Configuration {
```
I shadowed makeServerConf with an exported version but probably shouldn't keep it this way. I wanted to check if you think it is OK to export.

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -231,6 +242,74 @@ configuration.
+				checkErr("creating snapshot store", err)
+				dummyHost, err := makeDummyHost(clusterCfg)
+				checkErr("creating dummy host for snapshot store", err)
+				dummyTransport, err := libp2praft.NewLibp2pTransport(dummyHost, consensusCfg.NetworkTimeout)
```
I think you can use `hraft.InmemTransport` rather than going through the trouble of creating a libp2p one?

ZenGround0 commented:
@hsanjuan I am in agreement with you for the most part.  I've moved the version checking to the state module and moved the raft specific parts of migration to the consensus/raft module.  One place where I could use a little more advice is where you say "Instead of checking if the state needs a migration within the consensus component, you could do it in main.go, before you load anything."

I've been planning this out and if I try to validate the version of the state of existing raft snapshots before loading the consensus in main.go then I think I would end up needing to do a bunch of messy stuff like creating dummy raft components to call HasExistingState and traversing the raft data folder in search of snapshots files to load manually.

Would it be acceptable to check state validity after loading things, i.e. after calling NewCluster from main.go?  This could still be done from within main.go so that the only dependence introduced between consensus/raft and state/mapstate is in raft.Reset(mapstate) and would simplify the check.

Thanks!

hsanjuan commented on consensus/raft/raft.go:
```diff
@@ -427,6 +448,147 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	if err != nil {
+		return err
+	}
+	newStateBytes, err := encodeState(newState)
```
let this function receive an already encoded state, perhaps a https://godoc.org/github.com/hashicorp/raft#FSMSnapshot object, which is what the FSM.Snapshot() provides.

Also probably rename to `SnapshotReset`. `Reset` sounds like a thing which could be maybe used by `Consensus` for something less specific in the future.

ZenGround0 replied:
@hsanjuan Hmmm correct me if I'm wrong but if this function receives an encoded state that implies the caller (ipfs-cluster-service/main.go) will be creating a dummy libp2p-raft object, loading up its FSM with state and then calling Snapshot().  This is actually what I started with before refactoring after your first round of comments, and I think you were right then that that approach forces the main function to be too raft aware.  

I do get that we would rather keep the encode / decode functionality confined to libp2p raft.  I was thinking maybe we could export Encode and Decode?  Writing the first draft of this showed me that the current libp2p-raft interface makes you jump through hoops to get encodings of states but at the end of the day leaks them out anyway so we wouldn't be changing what users can do, just how easy they can do it.  

Let me know what you think or if I am missing something here.

ZenGround0 replied:
Agree with the name change 👍

hsanjuan replied:
Hmm, fair enough, you've convinced me. Let's export Encode and Decode!

You will need to make the change in libp2p-raft, do `gx publish -f` there, then take the hash and `gx update go-libp2p-raft <newhash>` here. Once everything is good we'll merge there, make a release, import a new final hash here. Let me know if you have problems with it.

hsanjuan commented on consensus/raft/raft.go:
```diff
@@ -427,6 +448,147 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+        return err
+}
+
+func encodeState(state interface{}) ([]byte, error) {
```
`encodeState` and `decodeState` should go away (see comment above about using libp2p-raft for this)

ZenGround0 replied:
I agree but am in favor of exporting them from libp2p raft, see my response above.

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -231,6 +233,57 @@ configuration.
+To succesfully run a migration of an entire cluster, shut down each peer without
+removal, migrate using this command, and restart every peer.
+`,
+			ArgsUsage: "<BACKUP-FILE-PATH> [RAFT-DATA-DIR]",
```
maybe downcase the args names ? For consistency with `ctl`...

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -290,6 +343,27 @@ func daemon(c *cli.Context) error {
 
 	state := mapstate.NewMapState()
+
+	r, err := raft.ExistingStateReader(consensusCfg) // Note direct dependence on raft here
```
This would be slightly different (See comment above). I think we can move most of this logic to a function too.

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -328,6 +402,17 @@ func daemon(c *cli.Context) error {
 	return nil
 }
 
+func cleanupRaft(raftDataDir string) error {
```
oh, this seems dup. Perhaps export from `raft`?

hsanjuan commented on sharness/t0050-service-migration.sh:
```diff
@@ -0,0 +1,33 @@
+    ipfs-cluster-ctl pin add "$cid" &> test4 && sleep 2 &&
+    kill -1 CLUSTER_D_PID && sleep 2 &&
+    backup-file=$( ls test-config/backups/ | head -n 1)
+    ipfs-cluster-service migration "test-config/backups/"$(backup-file) "test-config/ipfs-cluster-data"
```
Are we forgetting to check if this command is successful ? It would seem that, in case of failure, it will go unnoticed

hsanjuan replied:
if everything is not chained with `&&` and there are no returns, sharness will happily keep going forward.

hsanjuan commented on sharness/t0050-service-migration.sh:
```diff
@@ -0,0 +1,33 @@
+test_expect_success IPFS,CLUSTER,V1STATE "cluster-service loads v1 state correctly" '
+    cid=`docker exec ipfs sh -c "echo test | ipfs add -q"` 
+    kill -1 CLUSTER_D_PID &&
+    ipfs-cluster-service migration "test-config/v1State" "test-config/ipfs-cluster-data"
```
same as above

hsanjuan commented on state/interface.go:
```diff
@@ -27,4 +27,6 @@ type State interface {
 	// Restore restores a snapshot from a reader
 	Restore(r io.Reader) error
+	// Check whether version of state in reader has correct version
+	VersionOk(r io.Reader) bool
```
Probably it will be more versatile to have a `Version(r io.Reader) int` that tells us the version, and we can do comparison with mapstate.Version outside.

hsanjuan commented on util.go:
```diff
@@ -153,3 +161,32 @@ func containsPeer(list []peer.ID, peer peer.ID) bool {
 	return false
 }
+
+func BackupState(baseDir string, state state.State) error {
```
I think we can pass *Config here. This function can probably be improved with a "backup retain number" which will be another Config key and who knows what else. So more versatile to just receive the Config.

hsanjuan commented on sharness/t0050-service-migration.sh:
```diff
@@ -0,0 +1,33 @@
+#! /bin/sh
```
Sharness! 👍

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -290,6 +343,27 @@ func daemon(c *cli.Context) error {
+			}
+			if err == nil {
+				logger.Error("An updated backup of this state has been saved")
+				logger.Error("to .ipfs-cluster/backups.  To setup state for use")
```
actually, saved to baseDir/backups...

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -231,6 +233,57 @@ configuration.
 			Action: daemon,
 		},
+		{
+			Name:   "migration",
```
migrate? upgrade? or `state upgrade`? Perhaps we should leave room for other state commands (`state inspect`, `state clean`, `state downgrade`, `state export`)...

ZenGround0 replied:
I like it!

hsanjuan commented:
Heads up, I have published go-libp2p-raft 1.1.1 QmSrgyB4gDbaQnTCSaE9taRxhJkCUc9p5pYFqtXGyTaTbR. It has the exported functions. You should be able to `gx update go-libp2p-raft github.com/libp2p/go-libp2p-raft` to use it.

ZenGround0 commented:
Thanks @hsanjuan, I'll ping you when I'm ready for another round of review.

ZenGround0 commented:
@hsanjuan sharness is killing me today but apart from the failure I like where the code is at.  Could you give it another pass?  

If anything happens to jump out at you that I am doing wrong with Sharness, particularly the second test, I would be very grateful for any help.  These tests pass on my machine albeit with more delay than is required from other tests and with sharness.sh reporting something about Hangup Signals, but consistently fail on Travis.  I've ran the steps manually on a minikube cluster and migration always goes smoothly so I suspect the issue is related to cluster setup/teardown more than the state update.

hsanjuan commented on ipfs-cluster-service/main.go:
```diff
@@ -290,6 +315,11 @@ func daemon(c *cli.Context) error {
 	state := mapstate.NewMapState()
+
+	if needsUpdate(clusterCfg, consensusCfg) {
+		return errors.New("unsupported state version")
```
tell user what command to use or what to do in this case

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,91 @@
+
+
+func upgrade(c *cli.Context) error {
+	if c.NArg() < 1 || c.NArg() > 2 {
```
do the argument checking on the Action() and then call upgrade() with the right values

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,91 @@
+		return err
+	}
+	//Record peers of cluster                                                                  
+	var peers []peer.ID
```
if SnapshotReset does as explained above, it should be happy just with current peer ID. Also this is going to break with my PR since we are going to support multiple addresses per peer ID.

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,91 @@
+		return err
+	}
+	backupFilePath := c.Args().First()
+	var raftDataPath string
```
I'm not super sold on letting the user provide a custom raftDataPath with all the logic it adds and needing to carry it down 3 levels of functions. perhaps just set consensusCfg.DataFolder to the value of raftDataPath?

 Note that consensusCfg.DataFolder is normally empty and that in that case, baseDir+defaultDataSubfolder should be used. That's mostly for the raft component's functions.

ZenGround0 replied:
Thanks, given the above info I'm happy using the default or config value.

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,91 @@
+
+func needsUpdate(cfg *ipfscluster.Config, cCfg *raft.Config) bool {
+	state := mapstate.NewMapState()
+	r, err := raft.ExistingStateReader(cCfg) // Note direct dependence on raft here
```
following from comments above, this would be like `validSnap, err := raft.LastState(ccfg, &state)

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,91 @@
+func needsUpdate(cfg *ipfscluster.Config, cCfg *raft.Config) bool {
+	state := mapstate.NewMapState()
+	r, err := raft.ExistingStateReader(cCfg) // Note direct dependence on raft here
+	if err == nil { //err != nil no snapshots so skip check
```
if validSnap == false && err == nil -> no snapshots / nothing to do
if validSnap == true && err == nil -> do GetVersion check and maybe tell user migration is needed
if validSnap == true && err != nil -> could not deserialize state from snapshot, migration very probably needed or snapshot corrupted
if validSnap == false && err != nil -> some error before trying to read snapshots

hsanjuan commented:
@ZenGround0 we have one problem here though, we don't snapshot anymore on raft.Shutdown.

This had problems when a node had been removed from the cluster (snapshotting would hang). In that situation we cannot check for peers, we cannot snapshot, we can only shutdown. Since we cannot request the peerset, it is hard to figure out that we have been removed (if we wanted to skip snapshotting on that case).

This means that:

* there might be a snapshot, but it does not have the latest state because there are some newer entries in the log
* there might not be a snapshot at all, only entries in the log

Either we find a way to raft.Snapshot() on every shutdown or the only other approach I see is to get inspiration from https://github.com/hashicorp/raft/blob/master/api.go#L307 and:

* Try to load last snapshot
* Load the logstore
* Restore snapshot with libp2p-raft.FSM (this does de-serialization)
* Run fsm.Apply for all missing log entries between the snapshotIndex and the log last index
* return (the state should be the last state then)

---

Maybe we are focusing this wrongly. We save a backup on every shutdown. We could directly load the last backup against the current mapstate and check the version. If the last backup has not the right version, then we suggest a migration from it. Then we don't have to load snapshots at all and we just have to make the migrate part (close to what it is now).

On the other side, our backups are inefficient and the snapshot store pretty much does exactly the same. Maybe it would also help if backup format for mapstate is the same as libp2p-raft-snapshot, so it's easier to convert a snapshot to a backup, and backwards, but it's also tying mapstate and raft together.

I'm going to keep giving some thought to this. My feeling is that having a "soft" RecoverCluster which takes the last snapshot or makes a new one, takes the log, commits everything on it and saves it as a newer snapshot, is a useful thing to have. But let's test if we can snapshot right before raft.Shutdown() again.

hsanjuan commented:
Our main problem here is that we have a mapstate implementation which does not know how consensus->raft->libp2p raft stores states, and a consensus implementation that does not know if the mapstate implementation has a new format.

This https://github.com/libp2p/go-libp2p-raft/pull/29 introduces the possibility for MapState to specify how it wishes to be serialized.

With this, I'm going to take another try at the workflow, assuming snapshots are saved in a format that mapstate can understand directly because it implements Marshal/Unmarhsal itself.

* Let's find a way to save a snapshot on shutdown. I think it might work with `ShutdownOnLeave` set to true (this way attempts will fail when peer has been removed, but not hang at least). We assume the snapshot is the final known state.
* To check if a migration is needed:
  * main.go calls LastSnapshot() which returns the raw snapshot bytes of the last raft snapshot.
  * main.go creates a temp mapstate and calls mapstate.VersionFromBytes(bytes) which just deserializes into a `{ Version int }` object (we assume all our state formats will have a Version key) and tells us the version/ Deserializing the whole thing might fail if the new state format is not backwards compatible. I think it's nicer to just check the version part.
 * if the version from the read snapshot does not match the last Mapstate.Version, a migration is needed. We tell the users that they have Version X but current cluster uses Y.
* To do the migration:
  * Mapstate.Restore from the last snapshot's bytes
  * raft.ResetSnapshot(newstate) - like now, but adding a new snapshot based on the last one - same peers, same indexes, just different contents)

Note that we can forget about backups, since fully rely on raft snapshot store. And we should not cleanRaft() so harshly anymore either on some shutdowns, we should just rename the ipfs-cluster-data folder to something else.

Does this make sense? It's not so different from existing approach, except the snapshot format is now in control of mapstate. and simplifies `LastSnapshot` a bit.

Let's have a chat about it on monday to make sure we are on the same page?

Sorry for the loong notes, this is a very tricky part in cluster and it should work very well so I want a very clean approach (and something that we can cleanly split in the future if we need to).

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -427,9 +434,88 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	if err != nil {
+		return nil, err
+	}
+	stateBytes, err := json.Marshal(state)
```
Thank you @hsanjuan for explaining.  To check my understanding, when you say this code is "using insider information" are you referring to this code importing the `consensus.State` type, or my implicit assumption that a consensus.State could be json-marshalled into bytes that themselves will work out to be json-unmarshalled into a valid mapState?  I understand why the second thing is a problem, is the first thing (importing the consensus module)  also a problem?

hsanjuan replied:
> my implicit assumption that a consensus.State could be json-marshalled into bytes that themselves will work out to be json-unmarshalled into a valid mapState

That's the problem. Importing the consensus module is not

ZenGround0 commented:
Thank you for the explanations @hsanjuan.  Some thoughts:
Overall I like it.  I am in full agreement about being less harsh on old raft data, giving mapstate (un)marshal powers, and incorporating info about old snapshots into new ones.

One thing we lose by making raft snapshots the one-and-only backup is the readability and ease of use of the json backups, which was one of the original goals of issue #51.  I think the way to make this happen would be through another `state` sub-command.  For example `state backup` could store a json backup in a different folder by reading from the latest snapshot, and `state load` could create a raft snapshot from the json backup (It seems like it could be useful to specify an entire state at once, for example while testing).

I need to look into why saving snapshots on shutdown is a problem more closely, but your comments above indicate that it is not going to be possible in all circumstances.  Right now it sounds like you are suggesting considering any state changes that is not snapshotted due to bad conditions on shutdown as something cluster should ignore when migrating that peer's state.  Is this correct?  In the first long comment you proposed the idea of a "Soft Recover cluster" function that addressed the problem of failed shutdown snapshots, but this does not seem to be in the proposed workflow.  It seems to me this "soft recover" functionality could happen lazily/in memory in calls to LastSnapshot, or possibly be called from its own `state` subcommand.  It seems to be important and achievable.  Do you think it belongs in this PR?

One final thing to remind ourselves is that by deciding that cluster's state backups are the same as raft snapshots we might be further entrenching ourselves into a version of cluster that becomes harder to generalize to other consensus models.

I really appreciate the long notes.  I'm excited that this is coming together, thanks again for all the feedback!

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -427,9 +434,88 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	raftTerm        := uint64(1) // We reset history to the beginning
+	configIndex     := uint64(1) // We reset history to the beginning
+	srvCfg := makeServerConf(peers)
+	sink, err := snapshotStore.Create(raftSnapVersion, raftIndex, raftTerm, srvCfg, configIndex, dummyTransport)
```
Thanks for this, all great points

ZenGround0 commented:
@hsanjuan I am happy with this apart from the libp2p-raft blocker.  Let me know how it looks.  Thanks!

hsanjuan commented on consensus/raft/raft.go:
```diff
@@ -427,9 +435,134 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+
+// SnapshotMigrate saves a raft snapshot containing a migrated version of the
+// latest snapshot's state.  If no state yet exists there is nothing to migrate
+// and it returns an error.  Old snapshots are cleaned up.
```
`old raft data is backed up and replaced by the new snapshot` right?

ZenGround0 replied:
Yes, I'll clear this up!

hsanjuan commented on state/mapstate/map_state.go:
```diff
@@ -92,22 +97,70 @@ func (st *MapState) Snapshot(w io.Writer) error {
+	logger.Debugf("The interpreted version: %d", v)
+	if v != Version { // snapshot is out of date
+		st.Version = v
+		st.toRestore = bs[1:]
```
I find it a bit indirect to set `.toRestore` here. Is it possible to just return an error and then call:

`st.Restore(bs []byte)`. Or `st.Migrate(bs []byte)`

The "indirection" that bothers me is that it is not clear that Restore can only be called after a failed Unmarshal() attempt.

ZenGround0 replied:
Getting rid of this with the new approach suggested a few comments up 🎉

hsanjuan commented on cluster_config.go:
```diff
@@ -392,3 +395,33 @@ func clusterSecretToKey(secret []byte) (string, error) {
 }
+
+// BackupState backs up a state according to this configuration's options
+func (cfg *Config) BackupState(state state.State) error {
```
should we get rid if this and assume consensus backs up our state just fine?

ZenGround0 replied:
Right now the output is unused and getting rid of this function and its one call won't break anything.  I am neutral about this.  Removing reduces clutter, keeping it around could be useful for future additions to state commands.

hsanjuan commented on consensus/raft/raft.go:
```diff
@@ -427,9 +435,134 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	if err != nil {
+		logger.Debug("error decoding snapsot")
+		return err
+	}
```
the code here and up is very similar to LastState()

ZenGround0 replied:
Will be fixed while addressing previous comment

hsanjuan commented on consensus/raft/raft.go:
```diff
@@ -427,9 +435,134 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+// latest snapshot's state.  If no state yet exists there is nothing to migrate
+// and it returns an error.  Old snapshots are cleaned up.
+// Only call when Raft is shutdown.
+func SnapshotMigrate(cfg *Config, newState state.State) error {
```
Do you think it would make sense to do the migrate in main.go. I mean, instead of calling SnapshotMigrate() and let this (which is part of consensus) do everything, do instead:

* `consensus.LastState(cfg, st) []byte` to load the last snapshot
* `st.Restore([]byte)` to migrate it
* `consensus.SaveState(cfg,st)` to write a new snapshot

Or is there something strongly preventing this?

ZenGround0 replied:
Thanks for this.  For some reason I thought there was something strongly preventing this but there actually isn't anything.  I am pretty moving to this will address all of your other comments.

ZenGround0 replied:
This opens up a bit more the question of how unmarshal handles an old version of snapshot.  The current plan is to record the version, throw away the data and return no errors.  If you'd rather throw an error I am ok with that too.

hsanjuan commented on state/interface.go:
```diff
@@ -25,6 +25,14 @@ type State interface {
-	// Restore restores a snapshot from a reader
-	Restore(r io.Reader) error
+	// Restore restores an outdated state to the current version
+	Restore() error
```
If Restore() is called from main (see comment above), it would be called on the mapstate directly and maybe doesn't need to be part of the interface (same as Snapshot). I think also GetLatestVersion() might be ok to remove too.

ZenGround0 replied:
Hmm... fair enough, I am wondering though what kind of state would not want a restore method?  To me it makes sense to keep that one on the interface even if we don't need to.  I will be happy to remove GetLatestVersion() now that I am bringing more things to main.

hsanjuan replied:
Well, the interface is more, 'what are the requirements of the state components so it can be used by cluster'. If cluster does not use the Restore method... but yeah, it doesn't bother me much anyway.

hsanjuan commented on sharness/t0050-service-state-upgrade-from-current.sh:
```diff
@@ -0,0 +1,24 @@
+cleanup test_clean_cluster
+
+test_expect_success IPFS,CLUSTER "cluster-service state preserved by migration" '
+    cid=`docker exec ipfs sh -c "echo testing | ipfs add -q"` &&
```
These 3 first don't need to be &&-chained I guess, since whats below will fail if they didn't work

hsanjuan commented on sharness/t0050-service-state-upgrade-from-current.sh:
```diff
@@ -0,0 +1,24 @@
+    sleep 50 &&
+    ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid" &&
+    ipfs-cluster-ctl status "$cid" | grep -q -i "PINNED" &&
+    pkill -1 -f "ipfs-cluster-service --config test-config" && sleep 2
```
I have the feeling we should add a kill() start() functions to the lib so we can do this a bit more cleanly.

ZenGround0 replied:
Hmm yeah I agree with you.  I tried adapting the current clean_cluster lib function to take in a pid for this test, but I ran back into the problem that within sharness executable strings `$!` doesn't actually record the latest spawned process.  So `pkill` seems like the best I can do right now.  I am cool with adding a `start` and `kill` that uses `pkill`.

hsanjuan replied:
start() can save the pid and kill can re-use that, if both are in the lib/test_lib.sh file. it works like that now (it kills on clean up)

hsanjuan commented on sharness/t0051-service-state-upgrade-from-old.sh:
```diff
@@ -0,0 +1,35 @@
+# a snapshot of v1 state pinning "test" (it's easier than taking a new one each
+# time with the correct metadata). Upgrade state, check that the correct
+# pin cid is in the state
+test_expect_success IPFS,CLUSTER,V1STATE,JQ "cluster-service loads v1 state correctly" '
```
this must have been very hard to get working <3

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -427,9 +435,98 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+// SnapshotSave saves the provided state to a snapshot in the
+// raft data path.  Old raft data is backed up and replaced
+// by the new snapshot
+func SnapshotSave(cfg *Config, newState state.State, meta *hraft.SnapshotMeta) error {
```
It certainly could.  I was addressing your earlier comment that `SnapshotSave` unnecessarily duplicates a lot of `LastStateRaw`'s code.  If we don't pass in meta, `SnapshotSave` will duplicate most of `LastStateRaw`.

We could choose not to receive the meta return value from `LastStateRaw` in main, change `SnapshotSave` to not take in meta, and then call `LastStateRaw` this time receiving meta from within `SnapshotSave`.  It would still keep the meta return argument but meta wouldn't be exposed in main.  

I'm pretty neutral about any position in this tradeoff space so let me know what you prefer.

hsanjuan replied:
maybe make a private LastStateRaw-like function which provides the meta, and call from both LastStateRaw and SnapshotSave ?

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -427,9 +435,98 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+	// make a new raft snapshot
+	var raftSnapVersion hraft.SnapshotVersion
+	raftSnapVersion = 1 // As of hraft v1.0.0 this is always 1
+	raftIndex := meta.Index
```
SGTM, let me know your preference on my response above and I will handle this either when encountering a nil meta or when searching for a snapshot and finding none.

ZenGround0 commented on package.json:
```diff
@@ -21,9 +21,9 @@
     {
       "author": "hsanjuan",
-      "hash": "QmcjpDnmS6jGrYrTnuT4RDHUHduF97w2V8JuYs6eynbFg2",
+      "hash": "QmWKmeVQuGMKDeLC6BEYE6Xz6mSUKrzDZALfcdAaVpurtX",
```
Thanks for the gx publish!

ZenGround0 commented on sharness/t0050-service-state-upgrade-from-current.sh:
```diff
@@ -0,0 +1,24 @@
+    kill -1 $CLUSTER_D_PID && sleep 30 &&
+    ipfs-cluster-service --config "test-config" state upgrade &&
+    ipfs-cluster-service --config "test-config" >"$IPFS_OUTPUT" 2>&1 &
+    sleep 50 &&
```
I have had a lot of trouble with cluster starting up slowly the second time around and needing this long sleep.  It may just be my local setup so I will try it out.

ZenGround0 commented:
Currently Travis always fails on test t0051 on travis, however this test consistently passes on my machine.  It has been tough to see exactly what is going on because output is not carrying over on restart for some reason.  One possibility that comes to mind is that t0051 accesses  a new data folder through relative paths, maybe travis doesn't allow this?

hsanjuan commented on ipfs-cluster-service/migrate.go:
```diff
@@ -0,0 +1,78 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
yeah this workflow looks very nice and straightforward now :)

hsanjuan commented:
@ZenGround0 this option needs to be `true` very possibly to avoid snapshot hanging when a peer is removed:

https://github.com/ipfs/ipfs-cluster/blob/master/consensus/raft/config.go#L233

ZenGround0 commented on consensus/raft/raft.go:
```diff
@@ -427,9 +434,106 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+// latestSnapshot looks for the most recent raft snapshot stored at the
+// provided basedir.  It returns a boolean indicating if any snapshot is
+// readable, the snapshot's metadata, and a reader to the snapshot's bytes
+func latestSnapshot(raftDataFolder string) (bool, *hraft.SnapshotMeta, io.ReadCloser, error) {
```
Good catch, looks like we can use meta!=nil as the boolean flag.

ZenGround0 commented:
@hsanjuan, got it.  Should I work on changing the default in this PR?

ZenGround0 commented:
@hsanjuan I think this is finally ready.  Let me know if there is more to do here before I can merge.

hsanjuan commented:
Thanks @ZenGround0 ! Great job here, this was a very complex change and I'm very glad we found a nice approach to it.
